### ISSUE TITLE
Compound primary key

### ISSUE THREAD

**Issue by elifarley** [rethinkdb/rethinkdb#3875]
When importing a csv file with 2 or more columns that are unique, it would be nice to specify it like this:

```
db.table_create('mytable', primary_key=['a', 'b'])

rethinkdb import --pkey a,b
```

And later on...

```
r.table('mytable').get(['BRAZIL', 'MG'])
```

**Comment by elifarley** [comment]
Looks like someone else on IRC would also like it:

> Guest39519: I have a problem trying to perform what according to the documentation is valid, create a table with a primary key being an array. The documentation says: ”The data type of a primary key is usually a string (like a UUID) or a number, but it can also be a time, binary object, boolean or an array.”
> Guest39519:r.db(’test’).tableCreate('a', {primaryKey: ['x', 'y']})
> Guest39519:gives me ”Expected type STRING but found ARRAY”
> Guest39519>i mean, a doc { x: 1, y: 2 } will need to be { x: 1, y: 2, some_dummy: [1, 2] }, just to be able to use arrays as primary keys? this conflicts semantically with how compound indexes work.
> Codepope:You create a secondary compound index with indexCreate
> Guest39519:those are not unique, which i need, hence the compound pk :/
> Guest39519:This has to be a rather common issue for people, no? How else would you implement a many-to-many relation table without two-way relations, and with at least one of the way being a compound key to ensure unique-ness?

**Comment by danielmewes** [comment]
That's an interesting idea.
We might be able to support general deterministic functions for generating the primary key actually.

Related to https://github.com/rethinkdb/rethinkdb/issues/3046

You probably know this, but just to have it here in case other users are wondering:
Right now the way to use a compound primary key is to store an array into the primary key field. For CSV imports this is not possible directly, but you can first import your data into table `tmp`, and then copy it with the correct key into the final table `final`:
```js
r.tableCreate('final', {primaryKey: 'id'}); // You can use a different name for the key field if you like
r.table('final').insert(r.table('tmp').merge({id: [r.row('a'), r.row('b')]}))
```

**Comment by deontologician** [comment]
Would it be possible to implement this as sugar on top of an array primary key? @mlucy @dnaielmewes?

**Comment by danielmewes** [comment]
@deontologician Well we need to add support for some sort of primary key transformation first. Currently the primary key is always the value of an actual field, and some parts of our code make use of this property.

So we would need to support some sort of primary key function.

I'd suggest that we limit the set of functions one can use though, to avoid upgrading issues if some ReQL term changes (you can rebuild secondary indexes pretty easily without downtime, but you cannot just rebuild the primary index).

**Comment by nstogner** [comment]
Any progress on compound primary keys?

**Comment by danielmewes** [comment]
@nstogner We have been very busy working on some other important features, and unfortunately haven't gotten to this yet. At the moment you still need to generate the compound primary keys manually in your application code.

**Comment by nstogner** [comment]
Thanks for the speedy response!

**Comment by tve** [comment]
I'm interested in this too and reading the docs thought it was supported. At least that's how I read https://www.rethinkdb.com/api/javascript/table_create/ where it says:
> Using an array as a primary key creates a compound index; read the documentation on compound secondary indexes for more information, as it applies to primary keys as well.

Maybe I'm misunderstanding what that means...

**Comment by danielmewes** [comment]
@tve The difference is that the documentation talks about storing an array in the single primary key field, while this issue is about supporting a primary key that combines data from multiple fields.

If you store arrays in the primary key field, the primary index will indeed behave somewhat like a compound index when querying.

I think the wording in the documentation is a bit misleading here. I'll open a separate issue in our docs repository tracker. Thanks for pointing it out.

**Comment by brandonbloom** [comment]
I had the same confusion as @tve from reading those docs. The quoted sentence immediately follows the options documentation, so it seemed like it was talking about the `primaryKey` option, not the value put in to a primary key field later.

I'm extremely new to RethinkDB, but I stumbled across this issue because I'd like to create a junction table for a many-to-many relationship and was hoping to have a unique constraint on the pair of both IDs, plus only two indexes (leftId, rightId), rather than three (primaryKey, leftId, rightId).

**Comment by ivan-kleshnin** [comment]
> If you store arrays in the primary key field, the primary index will indeed behave somewhat like a compound index when querying.

Is this usable at all? I'm trying a `towns` table like this:

```js
[
  {
    "id": ["belarus",  "minsk"] ,
    "major": 1 ,
    "name":  "Минск"
  },

  {
    "id": ["russia",  "moscow"] ,
    "major": 1 ,
    "name":  "Москва" 
  },

 ...
]
```

But I can't figure out how to filter by a part of this key:

```js
r.db("test").table("towns").filter((t) => {
  // return t("id")[0] == "russia" // nope
  // return t("id")(0) == "russia" // nope
  // ???
})
```

**Comment by toddmedema** [comment]
Agreed. Even if compound key support was just syntax sugar that mapped id to be `<key A>.<key b>`, would really simplify the final code to define it on table creation & not have to worry about running the sugar on only certain tables that're using compound keys.

**Comment by tve** [comment]
Where support for compound primary keys would be really helpful is to alleviate marshaling and unmarshaling issues on the client side. Maybe this is only an issue for typed languages, but for example, in Go if I have:
```
type User struct {
  Company string `gorethink:"company"`
  Name    string `gorethink:"name"`
  Age     int    `gorethink:"age"`
}
```
and I want a compound primary key on company&name then I need to transform this into

```
type User struct {
  ID   []string `gorethink:"id"` // really: [company, name]
  Age int       `gorethink:"age"`
}
```
and I have to manually extract Company and Name from the ID array.
